File: src/main/java/com/jetbrains/idear/actions/ExecuteVoiceCommandAction.java
package com.jetbrains.idear.actions;

import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.actionSystem.CommonDataKeys;
import com.intellij.openapi.actionSystem.DataContext;
import com.intellij.openapi.actionSystem.DataKey;
import com.intellij.openapi.editor.Editor;
import com.jetbrains.idear.actions.recognition.ActionCallInfo;
import com.jetbrains.idear.actions.recognition.TextToActionConverter;

import java.util.logging.Logger;

public class ExecuteVoiceCommandAction extends ExecuteActionByCommandText {
    private static final Logger logger = Logger.getLogger(ExecuteVoiceCommandAction.class.getSimpleName());
    public final static DataKey<String> KEY = DataKey.create("Idear.VoiceCommand.Text");

    @Override
    public void actionPerformed(AnActionEvent e) {
        DataContext dataContext = e.getDataContext();
        Editor editor = CommonDataKeys.EDITOR.getData(dataContext);

        TextToActionConverter provider = new TextToActionConverter(e.getDataContext());
        ActionCallInfo info = provider.extractAction(e.getData(KEY));
        if (info != null) {
            invoke(editor, info);
        }
        else {
            logger.info("Command not recognized");
        }
    }

}

File: src/main/java/org/openasr/idear/actions/ExecuteVoiceCommandAction.kt
package org.openasr.idear.actions

import com.intellij.openapi.actionSystem.*
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.diagnostic.logger
import org.openasr.idear.actions.recognition.ActionRecognizerManager
import org.openasr.idear.ide.IdeService
import org.openasr.idear.nlp.NlpResultListener.Companion.NLP_RESULT_TOPIC

/**
 * Similar to ExecuteActionFromPredefinedText but uses the `VoiceCommand.Text` data attached to the invoking `AnActionEvent`
 */
object ExecuteVoiceCommandAction : ExecuteActionByCommandText() {
    private val log = logger<ExecuteVoiceCommandAction>()
    private val messageBus = ApplicationManager.getApplication().messageBus

    override fun actionPerformed(e: AnActionEvent) {
//        log.info("project: ${e.project}")
//        e.dataContext.getData()
//        val component = e.getData(PlatformCoreDataKeys.CONTEXT_COMPONENT)
//        log.info("component: $component")

        val provider = ActionRecognizerManager(e.dataContext)
        val info = provider.handleNlpRequest((e.inputEvent as SpeechEvent).nlpRequest)

        if (info != null) {
            messageBus.syncPublisher(NLP_RESULT_TOPIC).onFulfilled(info)

            if (!info.fullfilled) {
                val editor = IdeService.getEditor(e.dataContext)
                if (editor == null) {
                    log.info("Invoking outside of editor: ${info.actionId}")
                    IdeService.invokeAction(info.actionId)
                } else {
                    runInEditor(editor, info)
                }
            }
        } else {
            log.info("Command not recognized")
        }
    }
}
